cleanlab_studio/internal/enrichment_utils.py

```diff
@@ -37,83 +43,97 @@ def extract_df_subset(
     return subset_df
 
 
-def get_compiled_regex_list(
-    regex: Union[str, re.Pattern[str], List[re.Pattern[str]]]
-) -> List[re.Pattern[str]]:
-    """Compile the regex pattern(s) provided and return a list of compiled regex patterns."""
+def get_regex_match_or_replacement(
+    response: str, regex: Union[str, Replacement, List[Replacement]]
+) -> Optional[str]:
+    """Performs regex matching (if input is a string) or replacements (if input is a tuple) to the response
+    given the matching patterns and replacement strings.
+    """
+    # if the regex is a string, do matching
     if isinstance(regex, str):
-        return [re.compile(rf"{regex}")]
-    elif isinstance(regex, re.Pattern):
-        return [regex]
+        compiled_pattern = re.compile(regex)
+        pattern_match = compiled_pattern.match(response)
+        if pattern_match:
+            return pattern_match.group(1)
+        return None
+
+    # if the regex is a tuple (or list of tuples), do replacement
+    if isinstance(regex, tuple):
+        regex_list = [regex]
     elif isinstance(regex, list):
-        return regex
+        regex_list = regex
     else:
         raise ValidationError(
-            "Passed in regex can only be type one of: str, re.Pattern, or list of re.Pattern."
+            "Passed in regex has to be either a string (pattern to match), a tuple (pair of patterns to match and replace), "
+            "or a list of tuples (pairs of patterns to match and replace)."
         )
 
+    for replacement_pair in regex_list:
+        if isinstance(replacement_pair, str):
+            raise ValidationError("Only a single regex pattern can be used for matching.")
 
-def get_regex_match(
-    response: str, regex_list: List[re.Pattern[str]], disable_warnings: bool
-) -> Optional[str]:
-    """Extract the first match from the response using the provided regex patterns. Return first match if multiple exist.
-    Note: This function assumes the regex patterns each specify exactly 1 group that is the match group using ``'(<group>)'``."""
-    for regex_pattern in regex_list:
-        pattern_match = regex_pattern.match(response)
-        if pattern_match:
-            return pattern_match.group(1)
-    if not disable_warnings:
-        warnings.warn(
-            f"Your provided regex: {str(regex_list)} did not match a single LLM output across the dataset. This message is simply to inform you that the regex is not having any effect."
-        )
-    return None
+        if not isinstance(replacement_pair, tuple) or len(replacement_pair) != 2:
+            raise ValidationError(
+                "Every item of the regex list must be a tuple that contains 2 strings: "
+                "(the regex pattern to match, the string to replace the matched pattern with)"
+            )
 
+        compiled_pattern = re.compile(replacement_pair[0])
```

Comment:
```suggestion
        # check regex is well formatted
        compiled_pattern = re.compile(replacement_pair[0])
```
Below is the function I used, no worries adding it now but I though i'd share. It would have to be modified a bit for this code
```python3
def check_regex_well_formed(
    regex_list: List[Tuple[re.Pattern[str], str]]
) -> None:
    """Function checks the regex for any too complex patterns that would take unexpectedly long to compute"""
    complex_patterns = [
        (re.compile(r'\\\d'), "backreference"), # backreference pattern
        (re.compile(r'\(\?<=|\(\?<!|\(\?=|\(\?!'), "lookahead/lookbehind"), # lookahead and lookbehind patterns
    ]

    for regex_pattern, sub_text in regex_list:
        for complex_pattern, complex_pattern_name in complex_patterns:
            if complex_pattern.search(regex_pattern):
                raise ValueError(f"Regular expressions can not contain {complex_pattern_name}. Only simplified regular expressions are allowed." )
            if complex_pattern.search(re.compile(sub_text)):   # if regex is applied in-order then this is necessary
                raise ValueError(f"Substitute text for regular expressions can not contain {complex_pattern_name}. Only simplified text is allowed." )
```